consensus/istanbul/backend/announce.go

```diff
@@ -497,32 +523,66 @@ func (sb *Backend) handleAnnounceMsg(peer consensus.Peer, payload []byte) error
 		logger.Trace("Going to process an announce msg", "announce records", announceData.AnnounceRecords)
 		for _, announceRecord := range announceData.AnnounceRecords {
 			if announceRecord.DestAddress == sb.Address() {
-				// TODO: Decrypt the enodeURL using this validator's validator key after making changes to encrypt it
-				enodeUrl := string(announceRecord.EncryptedEnodeURL)
-				node, err := enode.ParseV4(enodeUrl)
+				enodeBytes, err := sb.decryptFn(accounts.Account{Address: sb.Address()}, announceRecord.EncryptedEnodeURL, nil, nil)
 				if err != nil {
-					logger.Error("Error in parsing enodeURL", "enodeUrl", enodeUrl)
+					sb.logger.Warn("Error decrypting endpoint", "err", err, "announceRecord.EncryptedEnodeURL", announceRecord.EncryptedEnodeURL)
 					return err
 				}
-
-				if err := sb.valEnodeTable.Upsert(map[common.Address]*vet.AddressEntry{msg.Address: {Node: node, Version: announceData.Version}}); err != nil {
-					logger.Warn("Error in upserting a valenode entry", "AnnounceData", announceData.String(), "error", err)
+				enodeURL := string(enodeBytes)
+				node, err := enode.ParseV4(enodeURL)
+				if err != nil {
+					logger.Error("Error parsing enodeURL", "enodeUrl", enodeURL)
 					return err
 				}
-
+				if err := sb.answerAnnounceMsg(msg.Address, node, announceData.Version); err != nil {
+					logger.Warn("Error answering an announce msg", "target node", node.URLv4(), "error", err)
+					return err
+				}
+				// If the announce was only intended for this node, do not regossip
+				if len(announceData.AnnounceRecords) == 1 {
```

Comment:
This could be an attack vector, because the difference in behavior between a validator for whom the announce is intended or unintended is different. To avoid revealing a node as a validator, it must act indistinguishable from one that is not.